### ISSUE TITLE
TimeStamp exception in valid timestamp

### ISSUE THREAD

**Issue by johnkeates** [awslabs/kinesis-agent-windows#13]
Using the latest KinesisTap release on IIS8 Logs I see failures to parse the lines:

`2020-10-22 00:00:00.9431 MyIISLogSource ERROR 2020-10-21 21:59:01 W3SVC1 MY_MACHINE 12.29.33.35 GET /version.html - 80 - 12.29.153.3 HTTP/1.1 - - - intservices.exp.companyinternal.net 200 0 0 1014 87 0

2020-10-22 00:00:00.9431 MyIISLogSource ERROR Failed to get time stamp in D:\Logs\MY_MACHINE\W3SVC1\iis-2020-10-21.log, 45270226: System.NullReferenceException: Object reference not set to an instance of an object.
@stacktrace_ref 144 12802 1089308128`

**Comment by dhhoang** [comment]
@johnkeates Can you post your configuration for this source?

**Comment by johnkeates** [comment]
@dhhoang This is the source definition:
```
{
            "Directory": "D:\\Logs\\IIS\\W3SVC1",
            "FileNameFilter": "*.log",
            "Id": "MyIISLogSource",
	    "TimeZoneKind": "UTC",
            "SourceType": "W3SVCLogSource"
}
```

**Comment by johnkeates** [comment]
Looking at an identical source with a lower volume of logs, that one does work with no errors. From the logs I would expect a full StackTrace but there doesn't seem to be one, NLog.xml on Debug level doesn't add anything either (but I suspected as much from looking at the sources).

I don't understand why it would not have a record instance at that point (that's what that instance pointer refers to, right?).

The log data is exactly the same as all other IIS logs I have, no difference in line endings or binary formatting or anything like that.

**Comment by dhhoang** [comment]
@johnkeates  Can you verify that in your log files, there's a header line that specifies the fields according to https://docs.microsoft.com/en-us/previous-versions/iis/6.0-sdk/ms525807(v=vs.90)#w3c-extended-log-file-format 

The header line usually starts with "#Fields:", for ex `#Fields: time c-ip cs-method cs-uri-stem sc-status cs-version `

**Comment by johnkeates** [comment]
@dhhoang yes, it exists, but it doesn't always seem to be at the top of the file

**Comment by dhhoang** [comment]
@johnkeates could you profile some sample lines with that line included?

**Comment by johnkeates** [comment]
Here is a sample from a log from today:

```
#Software: Microsoft Internet Information Services 8.5
#Version: 1.0
#Date: 2020-11-02 00:00:00
#Fields: date time s-sitename s-computername s-ip cs-method cs-uri-stem cs-uri-query s-port cs-username c-ip cs-version cs(User-Agent) cs(Cookie) cs(Referer) cs-host sc-status sc-substatus sc-win32-status sc-bytes cs-bytes time-taken

```

**Comment by dhhoang** [comment]
@johnkeates Does the "#Fields" line line repeat in your log? If so, does it repeat daily? 
The agent seems to have hit a case where it's not able to figure out the field mapping from the "#Fields" line if it's not at the start of the log file.

**Comment by johnkeates** [comment]
@dhhoang Yes, it does repeat in the logs. In the file used in the example it was on line 100 and on line 1149769.

**Comment by dhhoang** [comment]
@johnkeates  it looks like if the `#Fields` line is not at the top of the file, the agent is unable to figure out the mapping and hence unable to parse the subsequent lines.
Would you want to have the option to specify a default field mapping?

**Comment by johnkeates** [comment]
@dhhoang A default field mapping would be great; in theory the mapping of a file is known thread of time for the W3SVC log format anyway. There are multiple versions of IIS using slightly more or less fields, so being able to configure a default to match is a great method to solve this.

Does it still update the internally known mapping once it encounters a new set of #Fields in the log?

**Comment by dhhoang** [comment]
> Does it still update the internally known mapping once it encounters a new set of #Fields in the log?

I think it would make sense to update the mapping after reading that line.

**Comment by dhhoang** [comment]
The fix is in progress. I'll update this issue accordingly.

**Comment by johnkeates** [comment]
Thanks for the update!

**Comment by dhhoang** [comment]
@johnkeates Please give a new beta https://github.com/awslabs/kinesis-agent-windows/releases/tag/1.2.3.3 a try and see if it fixes your issue
